stellar-contract-env-host/src/host.rs

```diff
@@ -1441,8 +1488,54 @@ impl CheckedEnv for Host {
         i: RawVal,
         j: RawVal,
         l: RawVal,
-    ) -> Result<RawVal, HostError> {
-        todo!()
+    ) -> Result<Object, HostError> {
+        // offset in the host binary
+        let i: usize = u32::try_from(i).map_err(|_| {
+            HostError::WithStatus(
+                String::from("i must be u32"),
+                ScStatus::HostFunctionError(ScHostFnErrorCode::InputArgsWrongType),
+            )
+        })? as usize;
+        // offset in the guest memory
+        let j: u32 = j.try_into().map_err(|_| {
+            HostError::WithStatus(
+                String::from("j must be u32"),
+                ScStatus::HostFunctionError(ScHostFnErrorCode::InputArgsWrongType),
+            )
+        })?;
+        // length of the slice to copy
+        let l: usize = u32::try_from(l).map_err(|_| {
+            HostError::WithStatus(
+                String::from("l must be u32"),
+                ScStatus::HostFunctionError(ScHostFnErrorCode::InputArgsWrongType),
+            )
+        })? as usize;
+
+        let res = self.visit_obj(b, move |hv: &Vec<u8>| {
+            if i > u32::MAX as usize - l {
```

Comment:
I stared at this for a while wondering if it was safe, and I think it is _if_ all the earlier conversions are done safely. This might end up clearer if you did all the calculations in `u32`, then safely converted the final results to `usize`.